python/ray/data/_internal/lazy_block_list.py

```diff
@@ -101,7 +101,7 @@ def get_metadata(self, fetch_if_missing: bool = False) -> List[BlockMetadata]:
         """Get the metadata for all blocks."""
         if all(meta is not None for meta in self._cached_metadata):
             # Always return fetched metadata if we already have it.
-            metadata = self._cached_metadata
+            metadata = self._cached_metadata.copy()
```

Comment:
It somehow got mutated by DatasetStats when it creates the summary string.

Reply:
Hmm creating the summary string definitely shouldn't mutate the underlying metadata, we should probably fix that and add a test that ensures that it doesn't mutate it.

Reply:
It does here: https://sourcegraph.com/github.com/ray-project/ray@master/-/blob/python/ray/data/_internal/stats.py?L253

Reply:
@jianoaix Ah I see, we're overwriting the per-read-task metadata in the per-stage stats there, which is (accidentally) overwriting the metadata list at the `BlockList` level.

In that case, instead of copying the cached metadata in the `LazyBlockList` here which seems a bit brittle, I would vote for doing one of the following:

## Option 1: Don't Mutate
Ensure that the stats summary code does **not** [mutate the block metadata lists it receives](https://github.com/ray-project/ray/blob/46383028f8ea8554c68216a5fdebc0bc67cbdaaa/python/ray/data/_internal/stats.py#L252-L253):
```python
            if DatasetContext.get_current().block_splitting_enabled:
                self.stages["read"] = []
                for _, blocks_metadata in sorted(stats_map.items()):
                    self.stages["read"] += blocks_metadata
            else:
                self.stages["read"] = [None] * len(self.stages["read"])
                for i, metadata in stats_map.items():
                    self.stages["read"][i] = metadata[0]
```

## Option 2: Pass a Copy
Ensure that we always pass a block metadata list **copy** [when constructing a `DatasetStats` object](https://github.com/ray-project/ray/blob/46383028f8ea8554c68216a5fdebc0bc67cbdaaa/python/ray/data/_internal/lazy_block_list.py#L117), assuming that the list may be overwritten.
```python
        return DatasetStats(
            stages={"read": self.get_metadata(fetch_if_missing=False).copy()},
            parent=None,
            needs_stats_actor=True,
            stats_uuid=self._stats_uuid,
        )
```

I'm thinking that option (1) is a more sensible contract?

Reply:
Making a copy seems make a bit more sense to me than expecting caller not to mutate it, as it's not enforced by something like const reference.

Reply:
I vote for option 2 as well.

Option 1 has the issue when user calls `ds.stats()` and stats actor does not get the metadata from read tasks yet. Because read tasks makes RPC to stats actor for sending the metadata. For dynamic block splitting, I am fixing the code path in https://github.com/ray-project/ray/pull/30093 as well.

Reply:
Option 2 sounds good to me, then!

Reply:
Alright, updated the PR.